### PR TITLE
Displaying a notice for risky Cash App transactions

### LINKED ISSUE #3198: [Cash App] Display a warning for transactions above 2000 USD
See: https://href.li/?https://docs.stripe.com/payments/cash-app-pay#payment-limits-and-options

Although Cash App Pay has no transaction limits, it may decline a high-amount transaction (i.e., 2000 USD) depending on the customer's account and transaction history.

### PR DESCRIPTION
**Pull request by wjrosa**
Fixes #3198

## Changes proposed in this Pull Request:

See: https://href.li/?https://docs.stripe.com/payments/cash-app-pay#payment-limits-and-options

Although Cash App Pay has no transaction amount limits, it may decline a high-amount transaction (i.e., 2000 USD) depending on the customer account and transaction history. This PR adds this notice to the checkout page when the cart amount reaches the threshold.

Screenshot 2024-06-17 at 15 59 13
Screenshot 2024-06-17 at 15 59 30

## Testing instructions

- Checkout to this branch on your local environment (`tweak/display-notice-for-risky-transactions-on-cash-app`)
- Run `npm install`, `npm run build:webpack` and `npm run up`
- Connect your Stripe account (you must use a US account)
- Enable Cash App in both your store settings and the Stripe dashboard
- As a shopper, add some products to your cart (below 2000 USD)
- Confirm that no warnings are displayed
- Add more products, above the 2000 USD
- Confirm that you can see the new notice warning about the decline risk

---

-   [x] Covered with tests (or have a good reason not to test in description ☝️)
-   [x] Added changelog entry **in both** `changelog.txt` and `readme.txt` (or does not apply)
-   [ ] Tested on mobile (or does not apply)

**Post merge**

-   [ ] Added testing instructions to the [Release Testing Instructions wiki page](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions) (or does not apply)

### CODE REVIEW

**Review by diegocurbelo — CHANGES_REQUESTED** [review]
Thanks for working on this @wjrosa!

The message is displayed twice in the Shortcode Checkout

Screenshot 2024-06-13 at 5 00 58 PM

Also, can we show the message only when the customer chooses CashApp? and maybe closer or inside the payment selection box instead of a notice at the top?

**Comment by wjrosa** [comment]
> Thanks for working on this @wjrosa!
> 
> The message is displayed twice in the Shortcode Checkout
> 
> Screenshot 2024-06-13 at 5 00 58 PM
> 
> Also, can we show the message only when the customer chooses CashApp? and maybe closer or inside the payment selection box instead of a notice at the top?

Thanks for your review, Diego! I have updated the implementation per your suggestion. It is shown now above the payment selection box and only when Cash App is selected 👍 

Edit: I am still working on the shortcode version 👀 .
Edit 2: Now it is complete and ready for a new review.

**Comment by wjrosa** [comment]
Some updated preview images:
Screenshot 2024-06-17 at 15 59 13
Screenshot 2024-06-17 at 15 59 30

File: client/blocks/upe/upe-deferred-intent-creation/payment-processor.js [hunk 1]
```diff
@@ -249,8 +270,9 @@ const PaymentProcessor = ( {
 
 	useStripeLink( api, elements, paymentMethodsConfig );
 
-	const updatePaymentElementCompletionStatus = ( event ) => {
-		setIsPaymentElementComplete( event.complete );
+	const onChange = ( { value, complete } ) => {
```
Comment (diegocurbelo) [inline client/blocks/upe/upe-deferred-intent-creation/payment-processor.js]: Maybe we can use a more self-explanatory name for the callback function, like `onSelectedPaymentMethodChange` or similar.
```suggestion
	const onSelectedPaymentMethodChange = ( { value, complete } ) => {
```

File: client/blocks/upe/upe-deferred-intent-creation/payment-processor.js [hunk 2]
```diff
@@ -269,7 +291,7 @@ const PaymentProcessor = ( {
 			/>
 			<PaymentElement
 				options={ getStripeElementOptions() }
-				onChange={ updatePaymentElementCompletionStatus }
+				onChange={ onChange }
```
Comment (diegocurbelo) [inline client/blocks/upe/upe-deferred-intent-creation/payment-processor.js]: Same as above...  `onSelectedPaymentMethodChange` or similar
```suggestion
				onChange={ onSelectedPaymentMethodChange }
```

File: client/stripe-utils/cash-app-limit-notice-handler.js [hunk 3]
```diff
@@ -0,0 +1,85 @@
+import { __ } from '@wordpress/i18n';
+import { callWhenElementIsAvailable } from 'wcstripe/blocks/upe/call-when-element-is-available';
+
+/** The amount threshold for displaying the notice. */
+export const CashAppNoticeAmountThreshold = 2000;
```
Comment (diegocurbelo) [inline client/stripe-utils/cash-app-limit-notice-handler.js]: In other places we use all capital letters with underscores separating words for constants, like:
- https://github.com/woocommerce/woocommerce-gateway-stripe/blob/8a22592a15ec7690e18703a6dc9bcb86593ca4bb/client/blocks/credit-card/constants.js#L2
- https://github.com/woocommerce/woocommerce-gateway-stripe/blob/8a22592a15ec7690e18703a6dc9bcb86593ca4bb/client/blocks/payment-request/constants.js#L1

```suggestion
export const CASHAPP_NOTICE_AMOUNT_THRESHOLD = 2000;
```

File: client/stripe-utils/cash-app-limit-notice-handler.js [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+/** The class name for the limit notice element. */
+const LimitNoticeClassName = 'wc-block-checkout__payment-method-limit-notice';
```
Comment (diegocurbelo) [inline client/stripe-utils/cash-app-limit-notice-handler.js]:
```suggestion
const LIMIT_NOTICE_CLASSNAME = 'wc-block-checkout__payment-method-limit-notice';
```

File: client/stripe-utils/cash-app-limit-notice-handler.js [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+ * @param {string} wrapperElementSelector The selector for the wrapper element.
+ * @param {number} cartAmount The cart amount.
+ * @param {boolean} isBlockCheckout Whether the checkout form is a block checkout.
+ */
+export function maybeShowCashAppLimitNotice(
+	wrapperElementSelector,
+	cartAmount = 0,
+	isBlockCheckout = false
+) {
+	if ( cartAmount <= CashAppNoticeAmountThreshold ) {
```
Comment (diegocurbelo) [inline client/stripe-utils/cash-app-limit-notice-handler.js]:
```suggestion
	if ( cartAmount <= CASHAPP_NOTICE_AMOUNT_THRESHOLD ) {
```

File: client/stripe-utils/cash-app-limit-notice-handler.js [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+CashAppLimitNotice.textContent = __(
+	'Please note that, depending on your account and transaction history, Cash App Pay may reject your transaction due to its amount.'
+);
+CashAppLimitNotice.setAttribute( 'data-testid', 'cash-app-limit-notice' );
+
+/**
+ * Remove the Cash App limit notice from the checkout form.
+ */
+export function removeCashAppLimitNotice() {
+	const limitNotice = document.querySelector( '.' + LimitNoticeClassName );
```
Comment (diegocurbelo) [inline client/stripe-utils/cash-app-limit-notice-handler.js]:
```suggestion
	const limitNotice = document.querySelector( '.' + LIMIT_NOTICE_CLASSNAME );
```

File: client/stripe-utils/cash-app-limit-notice-handler.js [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+
+/**
+ * The Cash App limit notice element.
+ */
+export const CashAppLimitNotice = document.createElement( 'div' );
+CashAppLimitNotice.classList.add( 'woocommerce-info', LimitNoticeClassName );
+CashAppLimitNotice.textContent = __(
+	'Please note that, depending on your account and transaction history, Cash App Pay may reject your transaction due to its amount.'
+);
+CashAppLimitNotice.setAttribute( 'data-testid', 'cash-app-limit-notice' );
```
Comment (diegocurbelo) [inline client/stripe-utils/cash-app-limit-notice-handler.js]: In other places we use camelCase for variables (regardless if they were defined with `var`, `const` or `let`).
```suggestion
export const cashAppLimitNotice = document.createElement( 'div' );
cashAppLimitNotice.classList.add( 'woocommerce-info', LIMIT_NOTICE_CLASSNAME );
cashAppLimitNotice.textContent = __(
	'Please note that, depending on your account and transaction history, Cash App Pay may reject your transaction due to its amount.'
);
cashAppLimitNotice.setAttribute( 'data-testid', 'cash-app-limit-notice' );
```

File: client/stripe-utils/cash-app-limit-notice-handler.js [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+) {
+	const wrapperElement = document.querySelector( wrapperElementSelector );
+	if ( appendToElement ) {
+		wrapperElement.appendChild( CashAppLimitNotice );
+	} else {
+		wrapperElement.insertBefore(
+			CashAppLimitNotice,
+			wrapperElement.firstChild
+		);
+	}
```
Comment (diegocurbelo) [inline client/stripe-utils/cash-app-limit-notice-handler.js]:
```suggestion
	if ( appendToElement ) {
		wrapperElement.appendChild( cashAppLimitNotice );
	} else {
		wrapperElement.insertBefore(
			cashAppLimitNotice,
			wrapperElement.firstChild
		);
	}
```

**Review by diegocurbelo — CHANGES_REQUESTED** [review]
Thanks for the changes @wjrosa!... is working perfectly for both checkouts!

I added a few renaming suggestions to follow the code style used in (most) of the extension.

**Comment by wjrosa** [comment]
Cool, @diegocurbelo ! I have included all your commit suggestions and fixed some issues with them. I didn't pay attention to all those standards, but I got them now 🙌

**Review by diegocurbelo — APPROVED** [review]
Thanks for fixing those; we should have better linting checks... it might be a nice project for the next hack week.

**Pull request merged**